src/dense.jl

```diff
@@ -67,9 +67,11 @@ struct AutoEnzyme{M, A} <: AbstractADType
     mode::M
 end
 
+AutoEnzyme{A}(mode::M) where {M, A} = AutoEnzyme{M, A}(mode)
```

Comment:
Apart from the ConstructionBase stuff (on which I might have to trust you, and that's okay), my main concern with this PR is here. Usually, constructors with only a subset of type parameters consider type parameters in the right order. This modification (unlike the one on `AutoForwardDiff`) is in conflict with the general convention.

Given a type `AutoEnzyme{T1,T2}`, if there is a shortcut `AutoEnzyme{T}`, I would expect `T` to refer to `T1` and not `T2`. This would also be more coherent with the way that [`UnionAll` types work](https://docs.julialang.org/en/v1/devdocs/types/#UnionAll-types). Otherwise, the two following uses are not in agreement:
```julia
backend = AutoEnzyme{T}(stuff) # T is T2
backend isa AutoEnzyme{T} # T is T1
```
What do you think?

Reply:
That hack is also what I had in mind, and I think it is better if we consider #96, i.e. the fact that user-unfriendly constructors `AutoSomething{T1,T2}(arg1, arg2)` may become part of the public API.